core/src/main/java/ai/timefold/solver/core/impl/domain/variable/listener/support/VariableListenerSupport.java

```diff
@@ -156,6 +179,23 @@ private void processShadowVariableDescriptorWithListVariable(ShadowVariableDescr
         }
     }
 
+    private void processShadowVariableDescriptorWithBothBasicListVariable(
+            ShadowVariableDescriptor<Solution_> shadowVariableDescriptor,
+            ListVariableStateSupply<Solution_> listVariableStateSupply) {
+        if (shadowVariableDescriptor instanceof InverseRelationShadowVariableDescriptor<Solution_> inverseRelationShadowVariableDescriptor) {
+            // If the source type is a collection, then it is a list variable
+            if (inverseRelationShadowVariableDescriptor.isSourceVariableCollectionType()) {
```

Comment:
Isn't there a more direct way of checking for this? Doesn't the shadow var descriptor point to its list var descriptor, or something like that? This feels like you're checking for a side effect, and not for the root cause.